### PR TITLE
Message factory

### PR DESCRIPTION
**Pull request by KPostOffice**
Neat li'l function for creating custom messages.  This is useful for testing and for people outside our organization who want to use the messaging module without adding to our repository or for testing before we add new messages to the repo.

### CODE REVIEW

**Comment by pacospace** [comment]
> Neat li'l function for creating custom messages. This is useful for testing and for people outside our organization who want to use the messaging module without adding to our repository or for testing before we add new messages to the repo.

What about adding documentation in the README file? :)

File: thoth/messaging/message_factory.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        class MessageContents(faust.Record, serializer="json"):
+            """Class used to represent a contents of a missing-package message Kafka topic."""
+
+            pass
+
+        MessageContents.__annotations__ = message_contents
+
+        def __init__(self, num_partitions: int = 1, replication_factor: int = 1):
+            """Initialize hash-mismatch topic."""
+            super(HashMismatchMessage, self).__init__(
```
Comment (pacospace) [inline thoth/messaging/message_factory.py:47]: Why `HashMismatchMessage` here?

File: thoth/messaging/message_factory.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+def message_factory(
+    t_name: str,
+    message_contents: Tuple[str, Union[type, GenericMeta]],
+    num_partitions: int = 1,
+    replication_factor: int = 1
+):
+    """Create new message types dynamically."""
+    class NewMessage(MessageBase):
+        """Class used for Package Release events on Kafka topic."""
```
Comment (pacospace) [inline thoth/messaging/message_factory.py:34]: docstring to be modified?

File: thoth/messaging/message_factory.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+
+"""This is Thoth Messaging module for HashMismatchMessage."""
```
Comment (pacospace) [inline thoth/messaging/message_factory.py:19]: Isn't it for any message?

File: thoth/messaging/__init__.py [hunk 4]
```diff
@@ -23,6 +23,7 @@
 from .missing_version import MissingVersionMessage
 from .package_releases import PackageReleaseMessage
 from .advise_justification import AdviseJustificationMessage
+from .message_factory import message_factory
 
 __name__ = "thoth-messaging"
 __version__ = "0.3.2"
```
Comment (pacospace) [inline thoth/messaging/__init__.py:29]: It should be updated to `0.3.4`

File: thoth/messaging/message_factory.py (hunk 1 above)
Comment (KPostOffice) [reply thoth/messaging/message_factory.py:47]: Oops, good catch

File: README.rst [hunk 5]
```diff
@@ -4,3 +4,15 @@ Thoth Messaging
 This provides a library called `thoth-messaging
 <https://pypi.org/project/thoth-messaging>`_ used in project `Thoth
 <https://thoth-station.ninja>`_.  It is a basic module to encapsule all messaging (here it is Kafka via Faust) primitives.
+
+Message Factory
+###############
+For the purposes of `Thoth` we have a few of our custom messages defined within the messaging module.  However, we
+wanted to allow individuals to be able to use our module without having to add their own messages first.  For testing
+and development purposes you can use `message_factory(...)` which allows you to create arbitrary messages using
+`thoth-messaging` using the following syntax:
```
Comment (pacospace) [inline README.rst:13]: Isn't it the de facto way to create messages from now on if we have this generalization? All messages created until now should be removed actually and make sure we use the `message_factory` everywhere in our repos :) WDYT @KPostOffice?

Comment (KPostOffice) [reply README.rst:13]: Hmmm, I'm not sure. I kinda like having all of our messages in a central location but I understand why we'd get rid of it so the library becomes less "thoth specific"

Comment (fridex) [reply README.rst:13]: I think the idea behind having these messages as classes is similar to having classes in ORM world. One can easily verify the message content and keep messages consistent in the application.

Comment (KPostOffice) [reply README.rst:13]: So we leave it as is then?

Comment (fridex) [reply README.rst:13]: IMHO it will require some maintenance and library versioning, managing releases. It might be beneficial long term.

**Comment by goern** [comment]
is this still valid? can we merge it?

**Comment by KPostOffice** [comment]
@goern Yeah, I didn't realize that this never got merged. I don't know what the final consensus was.

**Comment by goern** [comment]
Let’s merge it

**Pull request merged**